products/pdt/pdtlisten/src/listener.rs

```diff
@@ -0,0 +1,87 @@
+use ethers::{
+    providers::{Http, Middleware, Provider},
+    types::{Block, Transaction, U64},
+};
+
+use anyhow::Result;
+use async_stream::try_stream;
+use serde_json::{to_value, Value};
+use tokio::time::{interval, Duration};
+use tokio_stream::Stream;
+
+async fn get_block_by_number(x: U64, provider: &Provider<Http>) -> Result<Block<Transaction>> {
+    println!("found block with number {:?}, getting block info", x);
+    fn serialize(v: impl serde::Serialize) -> Value {
+        to_value(v).unwrap()
+    }
+    let mut raw_block: Value = provider
+        .request("eth_getBlockByNumber", [serialize(x), serialize(true)])
+        .await?;
+    // ZIL-5328 means our nonce is only one byte instead of 8, which ethers
+    // is not happy about.
+    raw_block["nonce"] = serde_json::to_value("0x0000000000000000")?;
+
+    let mut block: Block<Transaction> = serde_json::from_value(raw_block)?;
+    while block.number.is_none() {
+        println!("{:?} is pending, looping", x);
+
+        // loop until the block is no longer pending
+        // the sleep duration is set arbitrarily.
+        tokio::time::sleep(Duration::from_millis(1000)).await;
+        raw_block = provider
+            .request("eth_getBlockByNumber", [serialize(x), serialize(true)])
+            .await?;
+        raw_block["nonce"] = serde_json::to_value("0x0000000000000000")?;
+
+        block = serde_json::from_value(raw_block)?;
+    }
+    println!("found block number {:?}", block.number);
+    // println!("{:#?}", block);
+    Ok(block)
+}
+
+/// Fetches the most recent block number and compares against `last_seen_block_number` and retrieves all blocks in between
+async fn get_block(
+    provider: &Provider<Http>,
+    last_seen_block_number: &mut Option<U64>,
+) -> Result<Vec<Block<Transaction>>> {
+    let block_number = provider.get_block_number().await?;
+
+    let last_seen_block_number_unwrap = if let Some(_block_number) = last_seen_block_number {
+        _block_number.clone()
+    } else {
+        // if does not know last_seen_block, assumes it was just the one before
```

Comment:
I'm not sure this makes sense? If I understand correctly, this means if you start PDT against a brand new network, without any data in the metadata table in BQ, it will only start syncing blocks from the current height minus 1?

Reply:
The idea is to use the archive data to populate all the existing data. And listen to only begin once archive has been imported. So in this case use current height and have historical blocks imported

Reply:
This is more for a fix to manage smaller synchronisation issues